sdk/core/core-lro/src/poller.ts

```diff
@@ -0,0 +1,163 @@
+    super(message);
+    this.name = "PollerStoppedError";
+    Object.setPrototypeOf(this, PollerStoppedError.prototype);
+  }
+}
+
+export class PollerCancelledError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = "PollerCancelledError";
+    Object.setPrototypeOf(this, PollerCancelledError.prototype);
+  }
+}
+
+export abstract class Poller<TState, TResult> {
+  private stopped: boolean = true;
+  private resolve?: (value?: TResult) => void;
+  private reject?: (error: PollerStoppedError | PollerCancelledError | Error) => void;
+  private pollOncePromise?: Promise<void>;
+  private cancelPromise?: Promise<void>;
+  private promise: Promise<TResult>;
+  private pollProgressCallbacks: PollProgressCallback<TState>[] = [];
+  protected operation: PollOperation<TState, TResult>;
+
+  constructor(operation: PollOperation<TState, TResult>) {
+    this.operation = operation;
+    this.promise = new Promise(
+      (
+        resolve: (result?: TResult) => void,
+        reject: (error: PollerStoppedError | PollerCancelledError | Error) => void
+      ) => {
+        this.resolve = resolve;
+        this.reject = reject;
+      }
+    );
+  }
+
+  protected abstract async delay(): Promise<void>;
+
+  private async startPolling(): Promise<void> {
+    if (this.stopped) {
+      this.stopped = false;
+    }
+    while (!this.isStopped() && !this.isDone()) {
+      await this.poll();
+      await this.delay();
+    }
+  }
+
+  private async pollOnce(options: { abortSignal?: AbortSignalLike } = {}): Promise<void> {
+    const state = this.getOperationState();
+    try {
+      if (!this.isDone()) {
+        this.operation = await this.operation.update({
+          abortSignal: options.abortSignal,
+          fireProgress: this.fireProgress.bind(this)
+        });
+        if (this.isDone() && this.resolve) {
+          this.resolve(state.result);
+        }
+      }
+    } catch (e) {
+      state.error = e;
+      if (this.reject) {
+        this.reject(e);
+      }
+      throw e;
+    }
+  }
+
+  private fireProgress(state: TState): void {
+    for (const callback of this.pollProgressCallbacks) {
+      callback(state);
+    }
+  }
+
+  private async cancelOnce(options: { abortSignal?: AbortSignal } = {}): Promise<void> {
+    this.operation = await this.operation.cancel(options);
+    if (this.reject) {
+      this.reject(new PollerCancelledError("Poller cancelled"));
+    }
+  }
+
+  public poll(options: { abortSignal?: AbortSignal } = {}): Promise<void> {
+    if (!this.pollOncePromise) {
+      this.pollOncePromise = this.pollOnce(options);
+      const clearPollOncePromise = (): void => {
+        this.pollOncePromise = undefined;
+      };
+      this.pollOncePromise.then(clearPollOncePromise, clearPollOncePromise);
+    }
+    return this.pollOncePromise;
+  }
+
+  public async pollUntilDone(): Promise<TResult> {
+    if (this.stopped) {
+      this.startPolling().catch(this.reject);
```

Comment:
does this do anything? startPolling doesn't seem to throw when stopped and so this.reject never gets called?

Reply:
this is to make sure the error is handled, since I was getting unhandled errors on the console

Reply:
hmm is this some kind of strictness thing to make sure you don't have an unhandled promise chain maybe?

Reply:
not sure! Let me call you so that you see it with me